packages/nimble-components/src/table-column/icon/tests/table-column-icon.spec.ts

```diff
@@ -0,0 +1,432 @@
+                            </${tableColumnIconTag}>
+                        </${tableTag}>`
+                );
+            }
+            it('is invalid with text mapping', async () => {
+                ({ element, connect, disconnect } = await setupInvalidMappings());
+                await connect();
+                await waitForUpdatesAsync();
+                const column = element.columns[0] as TableColumnIcon;
+                expect(column.checkValidity()).toBeFalse();
+                expect(column.validity.unsupportedMappingType).toBeTrue();
+            });
+        });
+
+        it('catches duplicate key values', async () => {
+            ({ element, connect, disconnect } = await setup([
+                { key: 'a', label: 'alpha', icon: 'nimble-icon-xmark' },
+                { key: 'a', label: 'alpha', icon: 'nimble-icon-xmark' }
+            ]));
+            await connect();
+            await waitForUpdatesAsync();
+            const column = element.columns[0] as TableColumnIcon;
+            expect(column.checkValidity()).toBeFalse();
+            expect(column.validity.duplicateMappingKey).toBeTrue();
+        });
+
+        it('catches equivalent numeric key values', async () => {
+            ({ element, connect, disconnect } = await setup(
+                [
+                    { key: '0', label: 'alpha', icon: 'nimble-icon-xmark' },
+                    { key: '0.0', label: 'alpha', icon: 'nimble-icon-xmark' }
+                ],
+                'number'
+            ));
+            await connect();
+            await waitForUpdatesAsync();
+            const column = element.columns[0] as TableColumnIcon;
+            expect(column.checkValidity()).toBeFalse();
+            expect(column.validity.duplicateMappingKey).toBeTrue();
+        });
+
+        it('catches missing key value on non-default', async () => {
+            ({ element, connect, disconnect } = await setup([
+                { label: 'alpha', icon: 'nimble-icon-xmark' }
+            ]));
+            await connect();
+            await waitForUpdatesAsync();
+            const column = element.columns[0] as TableColumnIcon;
+            expect(column.checkValidity()).toBeFalse();
+            expect(column.validity.missingKeyValue).toBeTrue();
+        });
+
+        it('allows missing key value on default', async () => {
+            ({ element, connect, disconnect } = await setup([
+                {
+                    label: 'alpha',
+                    icon: 'nimble-icon-xmark',
+                    defaultMapping: true
+                }
+            ]));
+            await connect();
+            await waitForUpdatesAsync();
+            const column = element.columns[0] as TableColumnIcon;
+            expect(column.checkValidity()).toBeTrue();
+            expect(column.validity.missingKeyValue).toBeFalse();
+        });
+
+        it('catches non-icon icon value', async () => {
```

Comment:
In the HLD I think we listed some more cases of invalid icons that are worth testing. For example, if it's a valid Nimble icon but it couldn't be resolved because the client forgot to import it.

I think your check in the validator will catch those cases already but we should test for them.

Reply:
I couldn't practically test for that case. Our test build includes all tests together, so any imports are available to all tests. The icon tests import every icon, so I can't really test an un-imported Nimble icon.

I went ahead and added a test for a completely bogus icon name. That was the only other test case I could think of.